File: android/app/src/main/java/com/runformcoach/runformcoachai/sensor/AudioCoachEngine.kt
package com.runformcoach.runformcoachai.sensor

import android.content.Context
import android.speech.tts.TextToSpeech
import android.speech.tts.UtteranceProgressListener
import android.util.Log
import java.util.Locale

// ── CoachPrompt ──────────────────────────────────────────────────────────────

/** 教练提示类别。 */
enum class CoachCategory {
    cadence,
    verticalOscillation,
    groundContactTime,
    trunkLean,
    general
}

/**
 * 由 [AudioCoachEngine] 生成的教练语音提示。
 *
 * @property text      要朗读的文本
 * @property language  语言代码 ("en", "zh", "nl")
 * @property priority  优先级 (0=信息, 1=建议, 2=警告)
 * @property category  触发该提示的指标类别
 */
data class CoachPrompt(
    val text: String,
    val language: String = "en",
    val priority: Int = 0,
    val category: CoachCategory = CoachCategory.general
)

// ── AudioCoachEngine ─────────────────────────────────────────────────────────

/**
 * 实时语音教练 —— 通过 Android [TextToSpeech] 播报跑姿提示。
 *
 * 由指标阈值触发 (步频太低/太高、垂直振幅过大、躯干倾角不当、触地时间过长)。
 * 强制最小提示间隔 (默认 15s) 避免频繁打扰。
 * 支持英文、简体中文、荷兰语。
 *
 * 用法:
 * ```
 * val coach = AudioCoachEngine(context, language = "en", minInterval = 15)
 * coach.evaluate(cadence = sample, gait = snapshot)
 * ```
 */
class AudioCoachEngine(
    private val context: Context,
    /** 语言代码 ("en", "zh", "nl")。默认 "en"。 */
    language: String = "en",
    /** 两次语音提示之间的最小秒数 (限制 5–60)。 */
    val minIntervalSeconds: Int = 15
) {
    companion object {
        private const val TAG = "AudioCoachEngine"
        private const val CADENCE_HISTORY_MAX = 30
        private const val EARLY_SESSION_SKIP_SEC = 10.0
    }

    // ── 公开属性 ──────────────────────────────────────────────────────────────

    /** 当前语言代码。 */
    var language: String = language
        private set

    /** 最小提示间隔 (秒)。 */
    val minInterval: Int = maxOf(5, minOf(60, minIntervalSeconds))

    /** 当前是否正在播报。 */
    var isSpeaking: Boolean = false
        private set

    /** 是否静音 (不输出语音)。 */
    var isMuted: Boolean = false

    /** 提示排队时的回调 (供 UI 显示)。 */
    var onPromptQueued: ((CoachPrompt) -> Unit)? = null

    // ── 内部状态 ──────────────────────────────────────────────────────────────

    private var tts: TextToSpeech? = null
    private var ttsInitialized: Boolean = false
    private var lastPromptTimeMillis: Long = 0L
    private val cadenceHistory = ArrayDeque<Double>()
    private var promptCount: Int = 0

    // ── 初始化 ────────────────────────────────────────────────────────────────

    init {
        initTTS()
    }

    private fun initTTS() {
        tts = TextToSpeech(context) { status ->
            ttsInitialized = (status == TextToSpeech.SUCCESS)
            if (ttsInitialized) {
                tts?.language = resolveLocale(language)
                Log.i(TAG, "TTS 初始化成功, 语言: ${tts?.language}")
            } else {
                Log.e(TAG, "TTS 初始化失败, status=$status")
            }
        }
        tts?.setOnUtteranceProgressListener(object : UtteranceProgressListener() {
            override fun onStart(utteranceId: String?) {
                isSpeaking = true
            }
            override fun onDone(utteranceId: String?) {
                isSpeaking = false
            }
            @Deprecated("Deprecated in Java")
            override fun onError(utteranceId: String?) {
                isSpeaking = false
                Log.w(TAG, "TTS 播报错误, utteranceId=$utteranceId")
            }
        })
    }

    // ── 公开 API ──────────────────────────────────────────────────────────────

    /**
     * 根据当前指标评估是否需要播报教练提示。
     *
     * @param cadence         当前步频样本 (可为 null)
     * @param gait            最新步态快照 (可为 null)
     * @param targetCadence   本次训练目标步频 (默认 170)
     * @param elapsedSeconds  已用时 (秒)
     */
    fun evaluate(
        cadence: CadenceSample?,
        gait: GaitSnapshot?,
        targetCadence: Double = 170.0,
        elapsedSeconds: Double = 0.0
    ) {
        // 强制间隔
        val now = System.currentTimeMillis()
        val intervalMs = minInterval * 1000L
        if (now - lastPromptTimeMillis < intervalMs) return

        // 跳过训练前 10 秒 (让指标稳定)
        if (elapsedSeconds < EARLY_SESSION_SKIP_SEC) return

        val prompts = mutableListOf<CoachPrompt>()

        // ── 步频教练 ────────────────────────────────────────────────────────

        if (cadence != null && cadence.confidence >= 0.3) {
            cadenceHistory.addLast(cadence.stepsPerMinute)
            while (cadenceHistory.size > CADENCE_HISTORY_MAX) {
                cadenceHistory.removeFirst()
            }

            val avgCadence = cadenceHistory.sum() / cadenceHistory.size
            val delta = avgCadence - targetCadence

            if (delta < -15 && cadenceHistory.size >= 5) {
                prompts.add(makeCadencePrompt(delta = delta, direction = "low"))
            } else if (delta > 15 && cadenceHistory.size >= 5) {
                prompts.add(makeCadencePrompt(delta = delta, direction = "high"))
            }
        }

        // ── 步态教练 ────────────────────────────────────────────────────────

        if (gait != null) {
            // 垂直振幅
            if (gait.verticalOscillationCm > 10.0) {
                prompts.add(makeGaitPrompt(
                    CoachCategory.verticalOscillation,
                    value = gait.verticalOscillationCm,
                    unit = "cm",
                    threshold = 10.0
                ))
            }

            // 触地时间
            if (gait.groundContactTimeMs > 300) {
                prompts.add(makeGaitPrompt(
                    CoachCategory.groundContactTime,
                    value = gait.groundContactTimeMs,
                    unit = "ms",
                    threshold = 300.0
                ))
            }

            // 躯干倾角
            if (kotlin.math.abs(gait.trunkLeanDegrees) > 12) {
                prompts.add(makeGaitPrompt(
                    CoachCategory.trunkLean,
                    value = gait.trunkLeanDegrees,
                    unit = "degrees",
                    threshold = 12.0
                ))
            }
        }

        // ── 取最高优先级播报 ────────────────────────────────────────────────

        if (prompts.isNotEmpty()) {
            val bestPrompt = prompts.maxByOrNull { it.priority } ?: return
            speakSync(bestPrompt)
        }
    }

    /**
     * 立即播报一条自定义提示 (不受间隔限制，但遵循静音)。
     */
    fun speak(prompt: CoachPrompt) {
        if (isMuted) return
        speakSync(prompt)
    }

    /** 停止正在进行的语音。 */
    fun stopSpeaking() {
        tts?.stop()
        isSpeaking = false
    }

    /** 切换教练语言。 */
    fun setLanguage(newLanguage: String) {
        language = newLanguage
        tts?.language = resolveLocale(newLanguage)
    }

    /** 释放 TTS 资源。 */
    fun shutdown() {
        tts?.stop()
        tts?.shutdown()
        tts = null
        ttsInitialized = false
    }

    // ── 内部播报 ──────────────────────────────────────────────────────────────

    private fun speakSync(prompt: CoachPrompt) {
        lastPromptTimeMillis = System.currentTimeMillis()
        promptCount++
        onPromptQueued?.invoke(prompt)

        if (isMuted) return

        // 如 TTS 未初始化则尝试重新初始化
        if (!ttsInitialized || tts == null) {
            initTTS()
            if (!ttsInitialized) return
        }

        val utteranceId = "coach_$promptCount"
        tts?.apply {
            // 切换语速和音调 (稍慢，运动时更清晰)
            setSpeechRate(0.52f)
            setPitch(0.95f)
            language = resolveLocale(prompt.language)

            // Android TTS API: >21 使用 speak(CharSequence, int, Bundle, String)
            if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.LOLLIPOP) {
                speak(prompt.text, TextToSpeech.QUEUE_FLUSH, null, utteranceId)
            } else {
                @Suppress("DEPRECATION")
                speak(prompt.text, TextToSpeech.QUEUE_FLUSH, null)
            }
        }
    }

    // ── 提示工厂方法 ─────────────────────────────────────────────────────────

    private fun makeCadencePrompt(delta: Double, direction: String): CoachPrompt {
        val absDelta = kotlin.math.abs(delta).toInt()
        val priority = if (absDelta > 25) 2 else 1

        val text = when (language) {
            "zh" -> if (direction == "low") {
                "步频偏低，加快节奏，小步快跑。"
            } else {
                "步频偏快，放慢节奏，加大步幅。"
            }
            "nl" -> if (direction == "low") {
                "Cadans te laag. Verhoog je pasfrequentie met kortere passen."
            } else {
                "Cadans te hoog. Verlaag je tempo en verleng je passen."
            }
            else -> if (direction == "low") {
                "Cadence is $absDelta steps low. Shorten your stride and pick up the tempo."
            } else {
                "Cadence is $absDelta steps high. Lengthen your stride and settle into a rhythm."
            }
        }

        return CoachPrompt(
            text = text,
            language = language,
            priority = priority,
            category = CoachCategory.cadence
        )
    }

    private fun makeGaitPrompt(
        category: CoachCategory,
        value: Double,
        unit: String,
        threshold: Double
    ): CoachPrompt {
        val text = when (language) {
            "zh" -> when (category) {
                CoachCategory.verticalOscillation ->
                    "垂直振幅偏高，收紧核心，减少上下跳动。"
                CoachCategory.groundContactTime ->
                    "触地时间偏长，加快脚步转换，提高步频。"
                CoachCategory.trunkLean ->
                    if (value > 0) "身体前倾过多，稍微挺直上身。"
                    else "身体后仰，稍微前倾利用重力。"
                else -> "调整跑姿。"
            }
            "nl" -> when (category) {
                CoachCategory.verticalOscillation ->
                    "Te veel verticale beweging. Span je core aan en loop efficiënter."
                CoachCategory.groundContactTime ->
                    "Grondcontact te lang. Verhoog je pasfrequentie voor een lichtere landing."
                CoachCategory.trunkLean ->
                    if (value > 0) "Je leunt te ver voorover. Richt je bovenlichaam iets op."
                    else "Je leunt achterover. Helling iets naar voren voor betere voortstuwing."
                else -> "Pas je houding aan."
            }
            else -> when (category) {
                CoachCategory.verticalOscillation ->
                    "Reduce bounce. Engage your core and land softly."
                CoachCategory.groundContactTime ->
                    "Ground contact too long. Quick, light steps."
                CoachCategory.trunkLean ->
                    if (value > 0) "Leaning too far forward. Straighten up slightly."
                    else "Leaning back. Tilt forward slightly from the ankles."
                else -> "Adjust your form."
            }
        }

        return CoachPrompt(
            text = text,
            language = language,
            priority = 1,
            category = category
        )
    }

    // ── Locale 解析 ───────────────────────────────────────────────────────────

    /**
     * 将语言代码映射到 Android [Locale]。
     * "zh" → Simplified Chinese, "nl" → Dutch, 否则 → English US。
     */
    private fun resolveLocale(lang: String): Locale {
        return when (lang) {
            "zh" -> Locale.SIMPLIFIED_CHINESE
            "nl" -> Locale("nl", "NL")
            else -> Locale.US
        }
    }
}

File: ios/RunFormCoachAI/AudioCoachEngine.swift
import AVFoundation
import Foundation
import os.log

// MARK: - AudioCoachEngine

/// Real-time voice coach that speaks form cues via iOS `AVSpeechSynthesizer`.
///
/// Triggered by metric thresholds (cadence too low/high, excessive vertical oscillation,
/// poor trunk lean, long ground contact time). Enforces minimum prompt interval to avoid
/// nagging. Supports English, Chinese (Simplified), and Dutch.
///
/// Usage:
/// ```swift
/// let coach = AudioCoachEngine(language: "en", minInterval: 15)
/// coach.evaluate(cadence: 155, gait: snapshot)
/// ```
public final class AudioCoachEngine: NSObject, @unchecked Sendable, AVSpeechSynthesizerDelegate {

    // MARK: - Public properties

    /// The language code currently in use.
    public private(set) var language: String

    /// Minimum seconds between voice prompts.
    public let minInterval: TimeInterval

    /// Whether the coach is currently speaking.
    public private(set) var isSpeaking: Bool = false

    /// Whether the coach is muted (no speech output).
    public var isMuted: Bool = false

    /// Callback invoked when a prompt is queued (for UI display).
    public var onPromptQueued: (@Sendable (CoachPrompt) -> Void)?

    // MARK: - Private

    private let synthesizer = AVSpeechSynthesizer()
    private let queue = DispatchQueue(label: "com.runformcoachai.audio")
    private var lastPromptTime: Date?
    private var cadenceHistory: [Double] = []
    private var promptCount: Int = 0

    // MARK: - Init

    /// - Parameters:
    ///   - language: Language code ("en", "zh-Hans", "nl"). Default "en".
    ///   - minInterval: Minimum seconds between prompts (default 15, clamped 5–60).
    public init(language: String = "en", minInterval: TimeInterval = 15) {
        self.language = language
        self.minInterval = max(5, min(60, minInterval))
        super.init()
        synthesizer.delegate = self
    }

    // MARK: - Public API

    /// Evaluate current metrics and decide whether to speak a coaching prompt.
    ///
    /// - Parameters:
    ///   - cadence: Current steps per minute.
    ///   - gait: Latest gait analysis snapshot.
    ///   - targetCadence: Target cadence for this session (default 170).
    ///   - elapsedSeconds: Total seconds elapsed in the session.
    public func evaluate(
        cadence: CadenceSample?,
        gait: GaitSnapshot?,
        targetCadence: Double = 170,
        elapsedSeconds: TimeInterval = 0
    ) {
        queue.async { [weak self] in
            self?.evaluateSync(cadence: cadence, gait: gait,
                               targetCadence: targetCadence,
                               elapsedSeconds: elapsedSeconds)
        }
    }

    /// Speak a custom prompt immediately (respects mute, not interval).
    /// - Parameter prompt: The prompt to speak.
    public func speak(prompt: CoachPrompt) {
        guard !isMuted else { return }
        queue.async { [weak self] in
            self?.speakSync(prompt)
        }
    }

    /// Stop any in-progress speech.
    public func stopSpeaking() {
        queue.async { [weak self] in
            self?.synthesizer.stopSpeaking(at: .immediate)
            self?.isSpeaking = false
        }
    }

    /// Change the coaching language.
    /// - Parameter language: New language code ("en", "zh-Hans", "nl").
    public func setLanguage(_ language: String) {
        queue.async { [weak self] in
            self?.language = language
        }
    }

    // MARK: - AVSpeechSynthesizerDelegate

    public func speechSynthesizer(_ synthesizer: AVSpeechSynthesizer,
                                   didStart utterance: AVSpeechUtterance) {
        isSpeaking = true
    }

    public func speechSynthesizer(_ synthesizer: AVSpeechSynthesizer,
                                   didFinish utterance: AVSpeechUtterance) {
        isSpeaking = false
    }

    public func speechSynthesizer(_ synthesizer: AVSpeechSynthesizer,
                                   didCancel utterance: AVSpeechUtterance) {
        isSpeaking = false
    }

    // MARK: - Private evaluation

    private func evaluateSync(
        cadence: CadenceSample?,
        gait: GaitSnapshot?,
        targetCadence: Double,
        elapsedSeconds: TimeInterval
    ) {
        // Enforce interval
        if let last = lastPromptTime {
            guard Date().timeIntervalSince(last) >= minInterval else { return }
        }

        // Skip early session (first 10s: let metrics stabilise)
        guard elapsedSeconds >= 10 else { return }

        var prompts: [CoachPrompt] = []

        // --- Cadence coaching ---
        if let cad = cadence, cad.confidence >= 0.3 {
            cadenceHistory.append(cad.stepsPerMinute)
            while cadenceHistory.count > 30 { cadenceHistory.removeFirst() }

            // Use rolling average to avoid single-sample noise
            let avgCadence = cadenceHistory.reduce(0, +) / Double(cadenceHistory.count)
            let delta = avgCadence - targetCadence

            if delta < -15 && cadenceHistory.count >= 5 {
                // Cadence too low
                prompts.append(makeCadencePrompt(delta: delta, direction: "low"))
            } else if delta > 15 && cadenceHistory.count >= 5 {
                // Cadence too high
                prompts.append(makeCadencePrompt(delta: delta, direction: "high"))
            }
        }

        // --- Gait coaching ---
        if let gait = gait {
            // Vertical oscillation
            if gait.verticalOscillationCm > 10.0 {
                prompts.append(makeGaitPrompt(.verticalOscillation,
                                              value: gait.verticalOscillationCm,
                                              unit: "cm",
                                              threshold: 10.0))
            }

            // Ground contact time
            if gait.groundContactTimeMs > 300 {
                prompts.append(makeGaitPrompt(.groundContactTime,
                                              value: gait.groundContactTimeMs,
                                              unit: "ms",
                                              threshold: 300))
            }

            // Trunk lean
            if abs(gait.trunkLeanDegrees) > 12 {
                prompts.append(makeGaitPrompt(.trunkLean,
                                              value: gait.trunkLeanDegrees,
                                              unit: "degrees",
                                              threshold: 12))
            }
        }

        // --- Speak highest-priority prompt only ---
        if let prompt = prompts.max(by: { $0.priority < $1.priority }) {
            speakSync(prompt)
        }
    }

    private func speakSync(_ prompt: CoachPrompt) {
        lastPromptTime = Date()
        promptCount += 1
        onPromptQueued?(prompt)

        guard !isMuted else { return }

        let utterance = AVSpeechUtterance(string: prompt.text)
        utterance.voice = AVSpeechSynthesisVoice(language: voiceCode(for: prompt.language))
        utterance.rate = 0.52  // slightly slower for clarity during exercise
        utterance.pitchMultiplier = 0.95
        utterance.volume = 0.9

        synthesizer.stopSpeaking(at: .word)
        synthesizer.speak(utterance)
    }

    // MARK: - Prompt factories

    private func makeCadencePrompt(delta: Double, direction: String) -> CoachPrompt {
        let absDelta = Int(abs(delta))
        let priority: Int = absDelta > 25 ? 2 : 1

        let text: String
        switch language {
        case "zh-Hans":
            if direction == "low" {
                text = "步频偏低，加快节奏，小步快跑。"
            } else {
                text = "步频偏快，放慢节奏，加大步幅。"
            }
        case "nl":
            if direction == "low" {
                text = "Cadans te laag. Verhoog je pasfrequentie met kortere passen."
            } else {
                text = "Cadans te hoog. Verlaag je tempo en verleng je passen."
            }
        default: // en
            if direction == "low" {
                text = "Cadence is \(absDelta) steps low. Shorten your stride and pick up the tempo."
            } else {
                text = "Cadence is \(absDelta) steps high. Lengthen your stride and settle into a rhythm."
            }
        }

        return CoachPrompt(text: text, language: language,
                           priority: priority, category: .cadence)
    }

    private func makeGaitPrompt(_ category: CoachPrompt.CoachCategory,
                                 value: Double, unit: String,
                                 threshold: Double) -> CoachPrompt {
        let text: String
        switch (language, category) {
        case ("zh-Hans", .verticalOscillation):
            text = "垂直振幅偏高，收紧核心，减少上下跳动。"
        case ("zh-Hans", .groundContactTime):
            text = "触地时间偏长，加快脚步转换，提高步频。"
        case ("zh-Hans", .trunkLean):
            text = value > 0
                ? "身体前倾过多，稍微挺直上身。"
                : "身体后仰，稍微前倾利用重力。"
        case ("nl", .verticalOscillation):
            text = "Te veel verticale beweging. Span je core aan en loop efficiënter."
        case ("nl", .groundContactTime):
            text = "Grondcontact te lang. Verhoog je pasfrequentie voor een lichtere landing."
        case ("nl", .trunkLean):
            text = value > 0
                ? "Je leunt te ver voorover. Richt je bovenlichaam iets op."
                : "Je leunt achterover. Helling iets naar voren voor betere voortstuwing."
        default: // en
            switch category {
            case .verticalOscillation:
                text = "Reduce bounce. Engage your core and land softly."
            case .groundContactTime:
                text = "Ground contact too long. Quick, light steps."
            case .trunkLean:
                text = value > 0
                    ? "Leaning too far forward. Straighten up slightly."
                    : "Leaning back. Tilt forward slightly from the ankles."
            default:
                text = "Adjust your form."
            }
        }

        return CoachPrompt(text: text, language: language,
                           priority: 1, category: category)
    }

    /// Map our language codes to BCP-47 codes AVSpeechSynthesizer expects.
    private func voiceCode(for language: String) -> String {
        switch language {
        case "zh-Hans": return "zh-CN"
        case "nl":      return "nl-NL"
        default:        return "en-US"
        }
    }
}
